API to get condensed tree of modified blocks between two txg

mynktl opened the pull request:
### Description
API to get condense tree of modified blocks between two txg
Changes in ZVOL ZAP update/read API regarding microzap

<!--- Describe your changes in detail -->
Added uzfs_txg_block_diff API, which will return condensed tree of changed blocks
between two txg.
In ZVOL ZAP update/read API, value changed to 64bit unsigned integer instead of char *.

### How Has This Been Tested?
Unit test are added to test this change with uzfs_test.
To test this change : 
 - ./cmd/uzfs_test/uzfs_test -T 3 -t 10 -n 5
 - ./cmd/uzfs_test/uzfs_test -T 4 -t 10

### Types of changes
- [x] New feature (non-breaking change which adds functionality)

### Checklist:
- [x] My code follows the ZFS on Linux code style requirements.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain `Signed-off-by`.
- [ ] Change has been approved by a ZFS on Linux member.

Commit:
- API to get condensed tree of modified blocks between two txg
- changes in zvol zap update/read API for microzap checks

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

Commit:
travis build fix

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

vishnuitta commented on lib/libzpool/uzfs_zap.c:
```diff
@@ -61,8 +73,8 @@ uzfs_update_zap_entries(void *zvol, const uzfs_zap_kv_t **array,
 		kv = array[i];
-		VERIFY0(zap_update(os, ZVOL_ZAP_OBJ, kv->key, 1, kv->size,
-		    kv->value, tx));
+		VERIFY0(zap_update(os, ZVOL_ZAP_OBJ, kv->key, kv->size, 1,
```
I understand our kv->size is 8 currently. But, its better to be (.., 1, kv->size,..) @mynktl incase if kv->size becomes greater than 8 later on. Same goes to below also.

mynktl replied:
I kept it this way to avoid fat zap upgrade.

vishnuitta commented on lib/libzpool/uzfs_zap.c:
```diff
@@ -50,6 +51,17 @@ uzfs_update_zap_entries(void *zvol, const uzfs_zap_kv_t **array,
 
+	/*
+	 * check if key length is greater than MZAP_NAME_LEN.
+	 * key with MZAP_NAME_LEN+ length will convert microzap
```
Thanks @mynktl for these changes, so that, it won't become fat zap. Comment need to be fixed at this line.

mynktl replied:
ok

vishnuitta commented on lib/libzpool/uzfs_mtree.c:
```diff
@@ -0,0 +1,251 @@
+};
+
+void
+add_to_mblktree(avl_tree_t *tree, uint64_t boffset, uint64_t blen)
```
why tree called as mblktree..? also, add comments at each logic.

mynktl replied:
it contains modified block's details, so i kept it as mblktree. I will add comments at each tree operation/logic.

vishnuitta replied:
add_to_txg_diff_tree

vishnuitta commented on lib/libzpool/uzfs_mtree.c:
```diff
@@ -0,0 +1,251 @@
+	snprintf(snapname, sizeof (snapname), "%s@%s%llu", zv->zv_name,
+	    TXG_DIFF_SNAPNAME, now);
+
+	error = dsl_pool_hold(snapname, FTAG, &dp);
```
why don't we use dmu_objset_own rather than dsl_pool_hold & dsl_dataset_hold?

mynktl replied:
here, we are traversing snapshot only, so we can avoid dmu_objset_own. Regarding snapshot destruction while we are traversing snapshot, we can use dsl_dataset_long_hold.

vishnuitta commented on cmd/uzfs_test/uzfs_txg_diff.c:
```diff
@@ -0,0 +1,320 @@
+} wblkinfo_t;
+
+int
+del_from_mblktree(avl_tree_t *tree, uint64_t offset, uint64_t len)
```
operations on tree can go into one file in lib/libzpool/

mynktl replied:
in production code, we are not using this API. So, i didn't keep it in lib/libzpool. Once we need this, we will move this to lib/libzpool.

vishnuitta reviewed (changes requested):
changes looks good @mynktl .. but, we need to change data structures and functions names. Also, some more unit testing needed to see whether the traversal is returning correct values.

vishnuitta commented on lib/libzpool/uzfs_mtree.c:
```diff
@@ -0,0 +1,251 @@
+
+
+int
+uzfs_txg_block_diff(zvol_state_t *zv, uint64_t start_txg, uint64_t end_txg,
```
uzfs_get_txg_diff_tree

vishnuitta commented on lib/libzpool/uzfs_mtree.c:
```diff
@@ -0,0 +1,251 @@
+
+#define	TXG_DIFF_SNAPNAME	"tsnap"
+
+struct diff_txg_blk {
```
uzfs_txg_diff_cb_args_t

vishnuitta commented on lib/libzpool/uzfs_mtree.c:
```diff
@@ -0,0 +1,251 @@
+}
+
+int
+uzfs_changed_block_cb(spa_t *spa, zilog_t *zillog, const blkptr_t *bp,
```
uzfs_txg_diff_cb

vishnuitta commented on lib/libzpool/uzfs_mtree.c:
```diff
@@ -0,0 +1,251 @@
+#define	TXG_DIFF_SNAPNAME	"tsnap"
+
+struct diff_txg_blk {
+	avl_tree_t *tree;
```
uzfs_txg_diff_tree

vishnuitta commented on lib/libzpool/uzfs_mtree.c:
```diff
@@ -0,0 +1,251 @@
+}
+
+static int
+zvol_blk_off_cmpr(const void *arg1, const void *arg2)
```
uzfs_txg_diff_tree_compare

vishnuitta commented on cmd/uzfs_test/uzfs_txg_diff.c:
```diff
@@ -0,0 +1,320 @@
+}
+
+void
+uzfs_zvol_txg_diff_blk_test(void *arg)
```
uzfs_txg_diff_test

Commit:
changes according to review comments

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

Commit:
resolving confilicts from zfs-0.7-release

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

vishnuitta commented on cmd/uzfs_test/uzfs_txg_diff.c:
```diff
@@ -166,16 +186,37 @@ uzfs_zvol_txg_diff_blk_test(void *arg)
+			/*
+			 * make sure offset is aligned to block size
+			 */
 			offset = ((blk_offset * blksz + block_size) /
```
offset can be just blk_offset * block_size

vishnuitta replied:
also, have a test case with less active size (probably for 30K or 50K blocks), so that, collisions/merges of tree can be verified

vishnuitta commented on cmd/uzfs_test/uzfs_txg_diff.c:
```diff
@@ -166,16 +186,37 @@ uzfs_zvol_txg_diff_blk_test(void *arg)
+					offset_matched = B_TRUE;
+					break;
+				}
+				if (temp_blk->offset + temp_blk->len ==
```
this condition is not required at all

vishnuitta commented on cmd/uzfs_test/uzfs_txg_diff.c:
```diff
@@ -166,16 +186,37 @@ uzfs_zvol_txg_diff_blk_test(void *arg)
 
 	while (i++ < test_iterations) {
 		count = 0;
+		max_io = MAX(uzfs_random(100), 5);
```
make it as 10K..

vishnuitta reviewed (approved):
changes looks good Mayank. Just take care of those minor comments.

Commit:
- removing fix io count from functionality test of
	modified blocks between two txg
- fix for active_size/vol_size argument parsing in uzfs_test

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

vishnuitta committed:
Merge branch 'zfs-0.7-release' into cstor49
